### PR TITLE
Core - Add in RESTUtil class for encoding namespace as URL variable

### PR DESCRIPTION
**Pull request by kbendick**
The REST Catalog uses the string value of a namespace in URLs.

Per the spec, the levels of a multi-level namespace are joined by the NULL byte character (`'\0'` or more correctly `'\u0000'`).

This adds a utility class which can:
1) Properly URL encode a `Namespace` for use as either a path or query parameter
2) Function for decoding the namespace string from a path variable into a Namespace
3) Some helpers for use with REST catalog configuration (function to get all config values that start with a specific prefix, and function to strip the trailing slash from a path if need be).

The items from (3) can likely be put into a shared `CatalogUtil` class as they occur in a few places, but to cause minimal disruption I'm declaring them separately. They can be merged together as a follow up.

This PR is to help unblock concurrent work on the REST catalog.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  public static String stripTrailingSlash(String path) {
+    if (path == null) {
+      return null;
+    }
+
+    return path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
+  }
+
+  public static Map<String, String> filterByPrefix(Map<String, String> properties, String prefix) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:50]: nit: maybe add some javadoc to make it clear what the purpose of this is

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    properties.forEach((key, value) -> {
+      if (key.startsWith(prefix)) {
+        result.put(key.substring(prefix.length()), value);
+      }
+    });
+
+    return result;
+  }
+
+  public static String asURLVariable(Namespace ns) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:62]: nit: maybe just `urlEncode(Namespace ns)` and `urlDecode(String encodedNs)`, since `Variable` feels kind of misleading

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        throw new UncheckedIOException(
+            String.format("Failed to URL encode namespace as UTF-8 encoding is not supported: %s", ns), e);
+      }
+    }
+
+    return NULL_JOINER.join(encodedLevels);
+  }
+
+  public static Namespace fromURLVariable(String encodedNs) {
+    Iterable<String> encodedLevels = NULL_SPLITTER.split(encodedNs);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:79]: nit: maybe add a `null` check for the `encodedNs`

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:62]: also I think it would make sense to add a `null` check for the `ns`

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  public static String asURLVariable(Namespace ns) {
+    String[] levels = ns.levels();
+    String[] encodedLevels = new String[levels.length];
+
+    for (int i = 0; i < levels.length; i++) {
+      try {
+        encodedLevels[i] = URLEncoder.encode(levels[i], StandardCharsets.UTF_8.name());
+      } catch (UnsupportedEncodingException e) {
+        throw new UncheckedIOException(
+            String.format("Failed to URL encode namespace as UTF-8 encoding is not supported: %s", ns), e);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:71]: nit: wdyt about adding the actual namespace name right in the middle? `Failed to URL encode namespace '%s'as UTF-8 encoding is not supported` 
I think this reads a bit more fluent to users running into this error msg, wdyt? Same for the decoding message below maybe

**Review by nastra — COMMENTED** [review]
overall LGTM, just a few nits/suggestions

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 4 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:71]: Yeah I agree that reads better. Thanks for the suggestion.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    public String parameterEncodedForUrlUsage;
+
+    LevelsAndURLEncodedString(String[] levels, String parameterEncodedForUrlUsage) {
+      this.levels = levels;
+      this.parameterEncodedForUrlUsage = parameterEncodedForUrlUsage;
+    }
+  }
+
+  @Test
+  public void testFilterAndRemovePrefix() {
```
[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java:43]: Is there a performance concern vs using the `@Parameterized` annotation that instantiates a new test class?

The performance overhead of using `@Parameterized` is why I didn't do that here. I'll look into this. Thank you!

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 1 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:50]: That's a good point.

The purpose of this function is to be able to filter out configuration values that are relevant to the catalog, like we do when we configure catalog properties via the spark session configurations. For example, in the jdbc catalog, we configure the following properties. 

```bash
    --conf spark.sql.catalog.my_catalog.jdbc.verifyServerCertificate=true \
    --conf spark.sql.catalog.my_catalog.jdbc.useSSL=true \
    --conf spark.sql.catalog.my_catalog.jdbc.user=admin \
    --conf spark.sql.catalog.my_catalog.jdbc.password=pass
```

So we'd pass in `spark.sql.catalog.my_catalog.jdbc.` (or just `jdbc.`) to get the configuration values by name.

The same thing could be done for the REST catalog.  So here's an example.

```
Map<String, String> props = ImmutableMap.of(spark.sql.catalog.my_catalog.rest.readTimeoutMillis, 5000);
Map<String, String> restCatalogProps = filterByPrefix(props, "spark.sql.catalog.my_catalog.rest.");
```

This function isn't nearly as important as the namespace parsing (which is what's needed more immediately), and arguably this could be achieved inline with guava Maps.filter etc, but we had a bug in the JDBC catalog a bit back because the functionality wasn't properly tested, so continuing to use a dedicated function that is tested seemed smarter. Eventually I'd like to merge them into a simpler `CatalogUtil` or something but right now I'm trying to keep changes more isolated.

Ultimately, we might get rid of this function (and I'd even happily take it out of this PR). I _have_ been using this function in my development process, and it's used in the Jdbc catalog too, but if we want to hold off on merging it I'd be fine with that.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 2 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java]: Added a null check to both (and renamed them). Also added tests for the null checl.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java:43]: no concern at all in terms of performance. I was rather thinking that the code could just look cleaner by separating input & expected result from the actual test code, since that makes reading the code easier. 

Note that in case you decide to use JUnit5 features here, you would have to add this to `build.gradle`:

```
project(':iceberg-core') {
  test {
    useJUnitPlatform()
  }
  ...
}
```

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:58]: I think having your explanation from this comment as a javadoc would be great in case you want to keep the function

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  public static String stripTrailingSlash(String path) {
+    if (path == null) {
+      return null;
+    }
+
+    return path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
+  }
+
+  /**
+   * Takes in a set of properties, and returns all properties begin with the designated prefix with the prefix
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:51]: nit: begin -> beginning

**Review by nastra — APPROVED** [review]

File: core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java:43]: Cool will check that out. Thank you.

[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java:43]: So for this specific test, I'm actually testing that the whole `input` map, when passed into the function, returns the `expected` map. Including that certain keys (which don't contain the prefix) are removed.

So I'm not sure if using `@ParameterizedTest` here really helps much. It will on `testTrailingSlash` as well as on the `testNamespaceUrlEncodeDecode` method below.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 6 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java]: Fixed and reworded.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  private RESTUtil() {
+  }
+
+  public static String stripTrailingSlash(String path) {
+    if (path == null) {
+      return null;
+    }
+
+    return path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:47]: The version in `LocationProviders` strips all trailing slashes:

```java

  private static String stripTrailingSlash(String path) {
    String result = path;
    while (result.endsWith("/")) {
      result = result.substring(0, result.length() - 1);
    }
    return result;
  }
```

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * Takes in a string representation of a namespace as used for a URL parameter
+   * and returns the corresponding namespace.
+   * <p>
+   * See also {@link #urlEncode} for generating correctly formatted URLs.
+   */
+  public static Namespace urlDecode(String encodedNs) {
+    Preconditions.checkArgument(encodedNs != null, "Invalid namespace: null");
+    Iterable<String> encodedLevels = NULL_SPLITTER.split(encodedNs);
+    String[] decodedLevels =
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:106]: Nit: It seems strange that the method above allocates a string and uses a for loop, while this one uses a stream and converts using `toArray`. I usually like to keep methods that work basically the same way consistent so they're easy to understand if you know what is happening in one. I'd probably reformat like the `urlEncode` method.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * Takes in a map, and returns a copy with entries with keys beginning with the designated prefix
+   * with the prefix removed. Any entries whose keys don't begin with the prefix are not returned.
+   * <p>
+   * This can be used to get a subset of the configuration related to the REST catalog, such as all
+   * properties from a prefix of `spark.sql.catalog.my_catalog.rest.` to get REST catalog specific properties
+   * from the spark configuration.
+   */
+  public static Map<String, String> filterByPrefix(Map<String, String> properties, String prefix) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTUtil.java:58]: This name is okay, but it doesn't quite describe what's happening because it is not just filtering by a key prefix, it is also removing the prefix. I wonder if something like `extractPrefixMap` or something would describe it better. I don't really have a good name for it, so it's probably okay as it is.

File: build.gradle [hunk 10]
```diff
@@ -203,6 +203,10 @@ project(':iceberg-common') {
 }
 
 project(':iceberg-core') {
+  test {
+    useJUnitPlatform()
+  }
```
Comment (rdblue) [inline build.gradle:208]: We have a lot of parameterized tests that don't require JUnit5. Here's an example: https://github.com/apache/iceberg/blob/master/api%2Fsrc%2Ftest%2Fjava%2Forg%2Fapache%2Ficeberg%2Fexpressions%2FTestExpressionSerialization.java#L31-L72

I generally prefer not to change the build in small PRs unless we need to. If we want to move to JUnit5, then we should take a look at that and discuss it on the dev list. My main concern there is the reordering of assert method arguments.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 7 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java]: Updated!

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 9 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java]: Changed it to extractPrefixMap. We can iterate on the name, but I also updated the doc to try to make it a bit more clear.

File: core/src/main/java/org/apache/iceberg/rest/RESTUtil.java (hunk 8 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/rest/RESTUtil.java]: Yeah I agree. I wasn't sure which I cared for more. I have updated to have the same format as `urlEncode`.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/rest/TestRESTUtil.java]: Based on another comment, we'll evaluate junit 5 introduction to core possibly in another PR vs in a utilities class PR (for example, in case people don't cherrypick this PR in a fork they maintain etc).

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @kbendick!
